sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -680,6 +681,14 @@ class Dataset[T] private[sql](
     joinWith(other, condition, "inner")
   }
 
+  /**
+   * Joins this [[Dataset]] returning a [[Tuple2]] for each pair using cartesian join
+   * Note: cartesian joins are very expensive without a filter that can be pushed down.
+   *
+   * @since 2.0.0
+   */
+  def joinWith[U](other: Dataset[U]): Dataset[(T, U)] = joinWith(other, lit(true), "inner")
```

Comment:
You don't actually need to use a true literal here, just construct a join without a condition.